### PR TITLE
[None][perf] Port remaining DeepSeek V4 optimizations to main

### PR DESCRIPTION
**Pull request by lfr-0531**
## Summary by CodeRabbit

* **New Features**
  * Added faster DeepSeek V4 sliding block table computation, including support for scratch-based workflows.
  * Added optional base64 handling for prompt token IDs in disaggregated chat serving.
  * Added smarter workload sizing for sparse attention paths to better handle long requests.

* **Bug Fixes**
  * Improved handling of invalid or out-of-range block mappings by returning safe fallback values.
  * Added stricter input validation for custom block table operations and token ID transport.

## Description

Port the four remaining optimizations from `feat/deepseek_v4` onto the latest GitHub `main`:

- Adapt DSA indexer prefill chunk size for long sequences.
- Bound transient indexer MQA logits memory with query tiling.
- Add opt-in context-side base64 int32 transport for disaggregated `prompt_token_ids`.
- Add the DeepSeek V4 block-table CUDA custom op and use it from the cache manager.

The cherry-picks were reconciled with current-main interfaces, including the current conversation protocol fields and KV cache manager v2 import path. The changes remain opt-in where applicable and add no new dependency.

Source commits: `acbcb93`, `d6d1b48`, `dfb0609`, and `b7a8a8f`.

## Test Coverage

- PASS: full C++/CUDA build and wheel generation with `scripts/build_wheel.py --benchmarks --use_ccache --cuda_architectures "90-real;100-real"`; the generated wheel was installed successfully.
- PASS: verified the installed `1.3.0rc21` package contains both DSA changes and that its built `libth_common.so` registers both DeepSeek V4 block-table operators.
- PASS: both block-table operators match the PyTorch reference for `max_blocks=17` and `max_blocks=256` on an idle B300 GPU (four cases total).
- PASS: `python -m py_compile` for all modified Python modules.
- PASS: changed-files pre-commit suite after rebasing onto the latest `main`.
- Environment note: direct pytest collection of `tests/unittest/_torch/custom_ops/test_deepseek_v4_block_table.py` is blocked before test execution by the local FlashInfer/CUTLASS Python mismatch (`cutlass.cute.nvgpu.OperandMajorMode` is unavailable). The equivalent operator/reference cases above pass by loading the installed custom-op library directly; CI will run the committed pytest coverage in its pinned environment.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

**Review by longlee0622 — APPROVED** [review]

File: tensorrt_llm/_torch/attention_backend/sparse/dsa.py [hunk 1]
```diff
@@ -2011,9 +2047,15 @@ def prepare_for_chunked_prefill(metadata: DSAtrtllmAttentionMetadata,
         else:
             # Use indexer's own chunking logic to prevent L^2 complexity of indexer MQA logits computation for long sequences.
             # This is only used when MLA chunked prefill is not enabled.
+            # Adapt chunk size to the batch's largest K_compressed (see
+            # select_indexer_chunk_size).
+            max_k_compressed = int(indexer_params.kv_lens[:num_contexts].max().
```
Comment (brb-nv) [inline tensorrt_llm/_torch/attention_backend/sparse/dsa.py:2052]: Non-blocking comment: Does it make sense to keep an `_cpu` in metadata instead of copying this over with a sync?

File: tensorrt_llm/_torch/attention_backend/sparse/deepseek_v4/cache_manager.py [hunk 2]
```diff
@@ -1330,14 +1258,13 @@ def compute_sliding_block_tables(
...
             self._device_num_contexts.fill_(num_contexts)
-            _compute_sliding_block_tables_with_scratch_compiled(
+            torch.ops.trtllm.deepseek_v4_compute_sliding_block_tables_with_scratch(
                 self._device_kv_cache_block_offsets_input,
                 device_copy_idx,
                 self._device_layer_attn_pool_ids,
                 self._device_valid_sliding_pool,
                 self._device_layer_attn_scales,
                 self._device_layer_offsets,
-                self._device_block_positions,
```
Comment (brb-nv) [inline tensorrt_llm/_torch/attention_backend/sparse/deepseek_v4/cache_manager.py:1340]: Is this the only use of `_device_block_positions`? If yes, can we remove the field entirely?

File: tensorrt_llm/_torch/attention_backend/sparse/dsa.py [hunk 3]
```diff
@@ -443,6 +456,29 @@ def split_prefill_chunks(
...
+# the largest compressed KV length in the batch. Entries are
+# (k_compressed_lower_bound_exclusive, chunk_size); >512K -> 8K, [256K, 512K]
+# -> 16K, otherwise unchanged.
+_INDEXER_CHUNK_SIZE_HEURISTIC = (
+    (512 * 1024, 8 * 1024),
+    (256 * 1024 - 1, 16 * 1024),
+)
+
+
+def select_indexer_chunk_size(configured_chunk_size: int,
```
Comment (brb-nv) [inline tensorrt_llm/_torch/attention_backend/sparse/dsa.py:470]: Nit: Can use an `_` prefix as this seems to be only for internal use within this module?

File: cpp/tensorrt_llm/kernels/deepseekV4BlockTable.cu [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+        }
+
+        if (!isValidTable)
+        {
+            output[outputOffset + blockId] = kBadPageIndex;
+            continue;
+        }
+
+        int32_t const base = blockOffsets[blockOffsetsOffset + blockId];
+        output[outputOffset + blockId] = base == kBadPageIndex ? kBadPageIndex : base * scale + layerOffset;
```
Comment (brb-nv) [inline cpp/tensorrt_llm/kernels/deepseekV4BlockTable.cu:307]: Nit: Use `applyScaleAndOffset()` instead?

**Review by brb-nv — APPROVED** [review]
Minor comments.

File: tensorrt_llm/_torch/attention_backend/sparse/dsa.py (hunk 1 above)
Comment (longlee0622) [reply tensorrt_llm/_torch/attention_backend/sparse/dsa.py:2052]: Thanks for the feedback. I'm resolving comments for now to merge the PRs. @lfr-0531 can review when he's back from vacation and address them in a follow-up PR.

**Review by JunyiXu-nv — APPROVED** [review]
LGTM for llm_api part.

**Review by juney-nvidia — APPROVED** [review]
Approved from API perspective
